### PR TITLE
[1] DatabricksRM: Use databricks-sdk to fetch token / workspace URL + several small improvements

### PR DESCRIPTION
**Pull request by dbczumar**
DatabricksRM: Use databricks-sdk to fetch token / workspace URL + several small improvements

### CODE REVIEW

File: dspy/retrieve/databricks_rm.py [hunk 1]
```diff
@@ -55,53 +47,72 @@ class DatabricksRM(dspy.Retrieve):
...
+        retriever_model = DatabricksRM(
+            databricks_index_name = "your_index_name",
+            docs_id_column_name="your_id_column",
+            text_column_name="your_text_column",
+            k=3
+        )
         dspy.settings.configure(lm=llm, rm=retriever_model)
         ```
 
-        Below is a code snippet that shows how to query the Databricks Direct Vector Access Index using the forward() function.
```
Comment (dbczumar) [inline dspy/retrieve/databricks_rm.py:63]: The `forward()` function is not what users call to retrieve documents with DatabricksRM

File: dspy/retrieve/databricks_rm.py [hunk 2]
```diff
@@ -55,53 +47,72 @@ class DatabricksRM(dspy.Retrieve):
...
+        Below is a code snippet that shows how to query the Databricks Direct Access Vector
+        Search Index using the DatabricksRM retrieval module:
+
         ```python
-        self.retrieve = DatabricksRM(query=[1, 2, 3], query_type = 'vector')
```
Comment (dbczumar) [inline dspy/retrieve/databricks_rm.py:65]: The preexisting implementation of `DatabricksRM` used an inconsistent / incorrect definition of `query_type`. In Databricks Vector Search, `query_type` defines the search algorithm (ANN or hybrid), *not* the type of input (text or vector). The type of input can be inferred from the Python value type. This PR updates the `query_type` argument to have consistent semantics with Databricks Vector Search while maintaining backwards compatible for the old argument values.

File: dspy/retrieve/databricks_rm.py [hunk 3]
```diff
@@ -55,53 +47,72 @@ class DatabricksRM(dspy.Retrieve):
...
+        if not _databricks_sdk_installed and (self.databricks_token, self.databricks_endpoint).count(None) > 0:
+            raise ValueError(
+                "To retrieve documents with Databricks Vector Search, you must install the"
+                " databricks-sdk Python library, supply the databricks_token and"
+                " databricks_endpoint parameters, or set the DATABRICKS_TOKEN and DATABRICKS_HOST"
+                " environment variables."
+            )
         self.databricks_index_name = databricks_index_name
-        self.columns = columns
+        self.columns = list({docs_id_column_name, text_column_name, *(columns or [])})
```
Comment (dbczumar) [inline dspy/retrieve/databricks_rm.py:115]: Automatically include the user-specified `docs_id_column` and `text_column` in the list of columns to retrieve from the vector search index (`self.columns`). Previously, users had to specify the docs ID and text columns in 2 places (the corresponding arg *and* the `columns` arg). If the user forgot to specify the docs ID / text columns in the `columns` arg, they would face a confusing error (ID / text column not found). This change fixes that problem.

File: dspy/retrieve/databricks_rm.py [hunk 4]
```diff
@@ -128,71 +139,73 @@ def _get_extra_columns(self, item: Dict[str, Any]) -> Dict[str, Any]:
         Returns:
             Dict[str, Any]: Search result column values, excluding the "text" and not "id" columns.
         """
-        extra_columns = {
-            k: v
-            for k, v in item.items()
-            if k not in [self.docs_id_column_name, self.text_column_name]
-        }
+        extra_columns = {k: v for k, v in item.items() if k not in [self.docs_id_column_name, self.text_column_name]}
```
Comment (dbczumar) [inline dspy/retrieve/databricks_rm.py:142]: This is just a formatting change from the precommit linter

File: dspy/retrieve/databricks_rm.py [hunk 5]
```diff
@@ -128,71 +139,73 @@ def _get_extra_columns(self, item: Dict[str, Any]) -> Dict[str, Any]:
...
-                },
+                **{"metadata": {k: v for k, v in json.loads(item["metadata"]).items() if k != "document_id"}},
             }
         return extra_columns
 
     def forward(
         self,
         query: Union[str, List[float]],
-        query_type: str = "text",
+        query_type: str = "ANN",
```
Comment (dbczumar) [inline dspy/retrieve/databricks_rm.py:153]: See https://github.com/stanfordnlp/dspy/pull/1564/files#r1780358443

File: dspy/retrieve/databricks_rm.py [hunk 6]
```diff
@@ -217,14 +228,117 @@ def forward(
...
+        sorted_docs = sorted(items, key=lambda x: x["score"], reverse=True)[: self.k]
 
         # Returning the prediction
         return Prediction(
             docs=[doc[self.text_column_name] for doc in sorted_docs],
-            doc_ids=[
-                self._extract_doc_ids(doc)
-                for doc in sorted_docs
-            ],
+            doc_ids=[self._extract_doc_ids(doc) for doc in sorted_docs],
```
Comment (dbczumar) [inline dspy/retrieve/databricks_rm.py:236]: This is just a formatting change from the linter

File: dspy/retrieve/databricks_rm.py [hunk 7]
```diff
@@ -217,14 +228,117 @@ def forward(
             items += [item]
 
         # Sorting results by score in descending order
-        sorted_docs = sorted(items, key=lambda x: x["score"], reverse=True)[:self.k]
+        sorted_docs = sorted(items, key=lambda x: x["score"], reverse=True)[: self.k]
```
Comment (dbczumar) [inline dspy/retrieve/databricks_rm.py:231]: This is just a formatting change from the linter

File: dspy/retrieve/databricks_rm.py [hunk 8]
```diff
@@ -204,9 +217,7 @@ def forward(
             )
 
         if self.text_column_name not in col_names:
-            raise Exception(
-                f"text_column_name: '{self.text_column_name}' is not in the index columns: \n {col_names}"
-            )
+            raise Exception(f"text_column_name: '{self.text_column_name}' is not in the index columns: \n {col_names}")
```
Comment (dbczumar) [inline dspy/retrieve/databricks_rm.py:220]: This is just a formatting change from the linter

File: dspy/retrieve/databricks_rm.py [hunk 9]
```diff
@@ -204,27 +217,128 @@ def forward(
...
 
         # Extracting the results
         items = []
-        for idx, data_row in enumerate(results["result"]["data_array"]):
```
Comment (dbczumar) [inline dspy/retrieve/databricks_rm.py:213]: `idx` was unused

File: docs/api/retrieval_model_clients/AzureCognitiveSearch.md [hunk 10]
```diff
@@ -1,5 +1,5 @@
 ---
-sidebar_position: 2
+sidebar_position: 3
```
Comment (dbczumar) [inline docs/api/retrieval_model_clients/AzureCognitiveSearch.md:2]: Place DatabricksRM second in the list of retrievers documented at https://dspy-docs.vercel.app/api/category/retrieval-model-clients

File: docs/api/retrieval_model_clients/ChromadbRM.md [hunk 11]
```diff
@@ -18,6 +18,7 @@ ChromadbRM(
 ```
 
 **Parameters:**
+
```
Comment (dbczumar) [inline docs/api/retrieval_model_clients/ChromadbRM.md:21]: These newlines were inserted by the pre-commit linter

**Comment by okhat** [comment]
@krypticmouse Any idea if the docs here are easy to fix before we merge?

**Comment by krypticmouse** [comment]
@okhat @dbczumar The issue is in docs/dspy-usecases.md file line 112, the link markdown is wrong:

https://github.com/stanfordnlp/dspy/blob/cdbea6eb9accf1f1d2cc3f8ce333502bddf3b8fb/docs/docs/dspy-usecases.md?plain=1#L112C1-L112C143

Change line 112 from this:
```
| **Langfuse** | [Link]([https://docs.langtrace.ai/supported-integrations/llm-frameworks/dspy](https://langfuse.com/docs/integrations/dspy)) |
```

To this:
```
| **Langfuse** | [Link](https://langfuse.com/docs/integrations/dspy) |
```

**Comment by krypticmouse** [comment]
If the build fails vercel won’t merge it so there would be no changes reflected in the existing website.

**Comment by arnavsinghvi11** [comment]
Thanks @dbczumar for the updates! All looks good. 

With the changes regarding https://github.com/stanfordnlp/dspy/pull/1564/files#r1780356012, can we update the [DSPy on Databricks blog post](https://www.databricks.com/blog/dspy-databricks) to reflect this? 

Specifically, how the retriever is configured and the retreived_results is called through DatabricksRM

```
retriever = DatabricksRM(
    databricks_index_name = "your_index_name",
    docs_id_column_name="id",
    text_column_name="field2",
    k=3
)
retrieved_results = DatabricksRM(query="Example query text", query_type="hybrid"))
```

**Comment by dbczumar** [comment]
Thanks @arnavsinghvi11, absolutely! I'll get that blog updated. I'll also file a follow-up PR with some test coverage, but I've QAed this PR manually

**Review by krypticmouse — APPROVED** [review]
Structurally LGTM!

**Pull request merged**